Updated with required admin scopes (#5057)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/terraform-enable-org-access/main/index.md

---
title: Enable Terraform access for your Okta org
excerpt: Create an Okta application and credentials that Terraform uses to manage the objects in your organization.
layout: Guides
---

> Create an Okta application and credentials that Terraform uses to manage the objects in your organization.

Author: Fixed

## Overview

Terraform is a tool that allows admins to automate your Okta org. Terraform requires an appropriate level of access to make changes to your org. You provide that access with an Okta API service app, which is the app type required for machine-to-machine communication.

An Okta API service app uses the OAuth 2.0 [Client Credentials authorization flow](/docs/guides/implement-oauth-for-okta-serviceapp/main/) to authorize Terraform. You create a public/private key pair as the client credentials for this flow. Okta stores the public key, and Terraform uses the private key in the configuration for access to your org.

In the Okta API service app that controls your Terraform integration, you specify the Okta objects that Terraform can access using API scopes. Scopes are like permissions to do a certain action on a type of resource. For example, the Terraform configuration in this article creates a test group in Okta, which requires the `okta.groups.manage` API scope.

> One way to determine the required scopes for the service app is to determine which actions an admin needs to perform the same kind of action. Next, find the corresponding scopes in the list of [OAuth Admin Management scopes](/docs/api/oauth2/#okta-admin-management) and add them to your service app. For example, adding a group in the Okta Admin console requires the `groups.manage` action. In the table, the corresponding scope is `okta.groups.manage`. Add the same scopes to your Terraform configuration.

Reviewer: I find this opening sentence a little confusing: "...which actions an admin needs to perform the same kind of action." Could it mean this?
"...which permissions an admin needs to perform the same kind of action."

Author: Thanks for catching the issue. You are correct. The changes are in the next commit.

Separate from granting API scopes, you must assign admin permissions to the app through groups of permissions called admin roles.

> **Note:** See [Control Terraform access to Okta](/docs/guides/terraform-design-access-security).

## Create an API service app to manage Terraform access

Create an Okta service app that authorizes Terraform to update your org:

> 1. In the Admin Console, select **Applications** > **Applications**.

Author: Change made in next commit.

## Overview

Terraform is a tool that allows admins to automate your Okta org. Terraform requires an appropriate level of access to make changes to your org. You provide that access with an Okta API service app, which is the app type required for machine-to-machine communication.

An Okta API service app uses the OAuth 2.0 [Client Credentials authorization flow](/docs/guides/implement-oauth-for-okta-serviceapp/main/) to authorize Terraform. You create a public/private key pair as the client credentials for this flow. Okta stores the public key, and Terraform uses the private key in the configuration for access to your org.

In the Okta API service app that controls your Terraform integration, you specify the Okta objects that Terraform can access using API scopes. Scopes are like permissions to do a certain action on a type of resource. For example, the Terraform configuration in this article creates a test group in Okta, which requires the `okta.groups.manage` API scope.

> One way to determine the required scopes for the service app is to consider which ones an admin needs to perform the same kind of action. Next, find the corresponding scopes in the list of [OAuth Admin Management scopes](/docs/api/oauth2/#okta-admin-management) and add them to your service app. For example, adding a group in the Okta Admin console requires the `groups.manage`. In the table, the corresponding scope is `okta.groups.manage`.

Author: In next commit

> You'll also add any required scopes to your Terraform configuration.

Author: I have no idea. I'm only acting as a style updater etc.

Use a built-in admin role to assign permissions:

1. In the Admin Console, open the service app and select **Admin roles**.
1. Click **Edit assignments**.
1. Click **Add assignment**.
1. In the **Complete the assignment** section, for **Role** select **Organization Administrator*** or  **Superuser Admin**.
1. Click **Save Changes**.

For more information on custom roles, see [Custom admin roles](https://help.okta.com/okta_help.htm?type=oie&id=ext-cstm-admin-roles).

## Grant scopes

OAuth scopes define permissions for an external API client like Terraform. For a full reference, see [OAuth 2.0 Scopes](/docs/api/oauth2/#oauth-20-scopes).

> **Note:** Granting new scopes to a service app requires that the admin has Super Administrator permission.

1. Open the Okta API service app that you created in the previous section.
1. Click the **Okta API Scopes** tab.
1. Find the desired scope. For this example, find `okta.groups.manage`.
1. Click **Grant** for the desired scope.
1. For some scopes, Okta requests confirmation. To confirm, click **Grant Scope**.

Repeat these steps for any other desired scopes.

> > **Important**: When you add other resource types to your Terraform configuration, you must make several changes. Grant the new scopes to the app. Modify your Terraform provider setup to request the new scopes. Depending on your existing setup, you may need to add more admin roles to the app. See [Assign admin roles](#assign-admin-roles).

Author: Change to Caution

> > **Important**: When you add other resource types to your Terraform configuration, you must make several changes. Grant the new scopes to the app. Modify your Terraform provider setup to request the new scopes. Depending on your existing setup, you may need to add more admin roles to the app. See [Assign admin roles](#assign-admin-roles).

Author: ```suggestion
> **Caution**: When you add other resource types to your Terraform configuration, you must make several changes. Grant the new scopes to the app. Modify your Terraform provider setup to request the new scopes. Depending on your existing setup, you may need to add more admin roles to the app. See [Assign admin roles](#assign-admin-roles).
```

## Create access credentials

Use a public/private key pair to control access to the service app used by Terraform. Store the public key in the API service app, and use the private key in your Terraform configuration. You can use either Okta or an external tool to generate the public/private key pair.

> > **Important:** Choose an access method based on the level of security required for your service app.

Author: ```suggestion
> **Caution:** Choose an access method based on the level of security required for your service app.
```

### Use Okta to generate the keys

1. In the Admin Console, open the service app and select **General**.
1. In the **Client Credentials** section, click **Edit** to change the client authentication method.
1. Select **Public key / Private key** as the **Client authentication** method.
1. Click **Add key**.
1. In the Add a public key dialog, click **Generate new key**.
1. In the **Private key** section, click **PEM** to show the PEM format (not JSON) format of the private key.
1. Copy the PEM private key and save it as a `.key` file.

>    > **Important:** The private key appears in this dialog only once. Losing the private key requires generating a new key pair.

Author: ```suggestion
   > **Caution:** The private key appears in this dialog only once. Losing the private key requires generating a new key pair.
```

1. If you created multiple keys in this application, set the status of any public key currently used by Terraform to **Active**. Set the status of the other public keys to **Inactive**.
1. For a production deployment, securely store the private key in a Terraform secrets management system or other key management system.

### Use an external tool to generate the keys

Audit your external tool to generate a key pair for security. Follow organization best practices and audit any code beforehand to maintain org security. Never use an untrusted third-party web site to generate keys for production systems.

Check that the generated private key is in PKCS#1 format, which is the format required by the Okta Terraform provider. In that format, the file that contains the private key begins with `-----BEGIN RSA PRIVATE KEY-----`. If the key isn't in the right format, convert it to the correct format using the OpenSSL command line program:

1. In a terminal, go to the file path where you saved the original private key.
1. Run OpenSSL to convert the key. One of the following command lines should work depending on your operating system and version of OpenSSL:

>    `openssl rsa -in {ORIGINAL_PRIVATE_KEY} -out {CONVERTED_PRIVATE_KEY} -traditional`

Reviewer 2: I think you meant these to be a bulleted list?
```suggestion
   * `openssl rsa -in {ORIGINAL_PRIVATE_KEY} -out {CONVERTED_PRIVATE_KEY} -traditional`
```

>    `openssl rsa -in {ORIGINAL_PRIVATE_KEY} -out {CONVERTED_PRIVATE_KEY}`

Reviewer 2: ```suggestion
   * `openssl rsa -in {ORIGINAL_PRIVATE_KEY} -out {CONVERTED_PRIVATE_KEY}`
```

>    `openssl-rsa -in {ORIGINAL_PRIVATE_KEY} -out {CONVERTED_PRIVATE_KEY}`

Reviewer 2: ```suggestion
   * `openssl-rsa -in {ORIGINAL_PRIVATE_KEY} -out {CONVERTED_PRIVATE_KEY}`
```

   In this command:

>    * `ORIGINAL_PRIVATE_KEY`: Set this to the file path of the original key.

Reviewer 2: If you're going to repeat "Set this to" in each item, my preference would be to put the placeholder in the sentence as I think it reads slightly easier...
```suggestion
   * Set `{ORIGINAL_PRIVATE_KEY}` to the file path of the original key.
```

>    * `CONVERTED_PRIVATE_KEY`: Set this to the file path for the converted key.

Reviewer 2: ```suggestion
   * Set `{CONVERTED_PRIVATE_KEY}` to the file path for the converted key.
```

   In the previous code sample, add your values to the following fields:

   * `org_name`: Your Okta org name, which is the first part of your organization's Okta domain before `.okta.com`, `.oktapreview.com`, or `.okta-emea.com`. For example, if your Okta domain is `example-org.oktapreview.com`, the org name is `example-org`. Don't include the `-admin` suffix, which corresponds to your Admin Console URL. To confirm your Okta domain in the Admin Console, click your username in the upper-right corner and look for the text below your email address. An Okta Developer org has an org name with the form `dev-<number>`, such as `dev-000001`.

>    * `base_url`: Your Okta org domain. This is the end part of your domain URL: `.okta.com`, `.oktapreview.com`, or `.okta-emea.com`.

Reviewer 2: In the code example above, base_url is set to okta.com, but your text here implies it should have the preceding period - .okta.com 
```suggestion
   * `base_url`: Your Okta org domain. This is the end part of your domain URL: `okta.com`, `oktapreview.com`, or `okta-emea.com`.
```

   * `client_id`: The client ID of the API service app that you created. In the Admin Console, click **Applications**, then click the name of your API service app. In the app editor, click **General**. Under **Client Credentials**, copy the **Client ID** value.
   * `private_key`: Either the path to the private key file or the private key itself. Okta recommends storing the key in a separate location and using a secrets and encryption management system, such as HashiCorp Vault.

>    * `scopes`: The list of API scopes required by your Terraform configuration. The example below creates a group, which requires the `okta.groups.manage` scope.

Reviewer 2: ```suggestion
   * `scopes`: The list of scopes required by your Terraform configuration. The following example creates a group, which requires the `okta.groups.manage` scope.
```
